machine/models/Sender.py

```diff
@@ -0,0 +1,178 @@
+
+		Inputs:
+		    tau (float): Temperature to be used for Gumbel Softmax.
+		    hidden_state (torch.tensor, optional): The hidden state to start the decoding. (default=None)
+		    Shape [batch_size, hidden_size]. If None, batch_size=1.
+
+		Outputs:
+		    output_sequence (torch.tensor): The generated decoded sequences. Shape [batch_size, output_len+1]
+		    E.g. of a sequence at prediction time [sos_id, predicted_1, predicted_2,...., predicted_outputlen]
+		    sequence_lengths (torch.tensor): The lengths of all the sequences in the batch. Shape [batch_size]
+
+		"""
+
+	def __init__(self, vocab_size, output_len, embedding_size,
+				 hidden_size, sos_id, eos_id, rnn_cell='gru', greedy=False):
+		super().__init__(vocab_size, output_len, hidden_size,
+						 input_dropout_p=0, dropout_p=0,
+						 n_layers=1, rnn_cell=rnn_cell, relaxed=True)
+
+		self.output_len = output_len
+		self.embedding_size = embedding_size
+		self.sos_id = sos_id
+		self.eos_id = eos_id
+		self.greedy = greedy
+
+		self.embedding = nn.Parameter(torch.empty((self.vocab_size, self.embedding_size), dtype=torch.float32))
+		self.rnn = self.rnn_cell(self.embedding_size, self.hidden_size)
+		self.linear_out = nn.Linear(self.hidden_size, self.vocab_size)
+
+		self._reset_parameters()
+
+	def _reset_parameters(self):
+		nn.init.normal_(self.embedding, 0.0, 0.1)
+
+		nn.init.constant_(self.linear_out.weight, 0)
+		nn.init.constant_(self.linear_out.bias, 0)
+
+		nn.init.xavier_uniform_(self.rnn.weight_ih)
+		nn.init.orthogonal_(self.rnn.weight_hh)
+		nn.init.constant_(self.rnn.bias_ih, val=0)
+		# # cuDNN bias order: https://docs.nvidia.com/deeplearning/sdk/cudnn-developer-guide/index.html#cudnnRNNMode_t
+		# # add some positive bias for the forget gates [b_i, b_f, b_o, b_g] = [0, 1, 0, 0]
+		nn.init.constant_(self.rnn.bias_hh, val=0)
+		nn.init.constant_(self.rnn.bias_hh[self.hidden_size:2 * self.hidden_size], val=1)
+
+	def _init_state(self, hidden_state, rnn_type):
+		"""
+			Handles the initialization of the first hidden state of the decoder.
+			Hidden state + cell state in the case of an LSTM cell or
+			only hidden state in the case of a GRU cell.
+
+			Args:
+			    hidden_state (torch.tensor): The state to initialize the decoding with.
+			    rnn_type (type): Type of the rnn cell.
+
+			Returns:
+			    state: (h, c) if LSTM cell, h if GRU cell
+			    batch_size: Based on the given hidden_state if not None, 1 otherwise
+		"""
+
+		# h0
+		if hidden_state is None:
+			batch_size = 1
+			h = torch.zeros([batch_size, self.hidden_size], device=device)
+		else:
+			batch_size = hidden_state.shape[0]
+			h = hidden_state # batch_size, hidden_size
+		
+		# c0
+		if rnn_type is nn.LSTMCell:
+			c = torch.zeros([batch_size, self.hidden_size], device=device)
+
+			state = (h,c)
+		else:
+			state = h
+
+		return state, batch_size
+
+	def _calculate_seq_len(self, seq_lengths, token, initial_length, seq_pos, 
```

Comment:
What's the idea behind `_calculate_seq_len` ? 
It's really slow and we don't make use of the sequence lengths at any later point as far as I can tell.

Reply:
We will need it later to prevent the receiver from processing the whole sequence if the EOS token was sampled before reaching the max length

Reply:
Mmmh the compute time is still extremely significant. Running it with the `_calculate_seq_len` and without on gpu is almost a 5x computation time increase (10 seconds avg per epoch vs 50 secs avg per epoch).

Reply:
You also don't currently use it in the receiver #70 .

Reply:
Would you suggest having a flag for optionally computing the seq lengths?

Reply:
A flag wouldn't really fix the efficiency, but it's ok I've just implemented it using mask and preformance is no longer an issue. 
Try it with this code instead of the `enumerate` for loop:
```
        if is_discrete:
            mask = token == self.eos_id
        else:
            max_predicted, vocab_index = torch.max(token, dim=1)
            mask = (vocab_index == self.eos_id) * (max_predicted == 1.0)
        mask *= seq_lengths == initial_length
        seq_lengths[mask.nonzero()] = seq_pos + n_sos_symbols
```
This adds one avg second runtime per epoch on GPU vs the current version which adds 37secs on avg per epoch. 
I don't believe I have push access to your fork so you need to push it, but try it first and lmk what you think.